### PR TITLE
Require Node.js 8, add TypeScript definition

### LINKED ISSUE #3: Cancel pending promises on resolution
For example, if one passes three promises and the first resolves, `.cancel()` the other two if they have such a method on the returned Promise. There should be an option to control this behaviour.

I think the implementation of this could be challenging based on the fact that there's no easy way to synchronously check if a Promise has resolved (https://stackoverflow.com/q/30564053).

Refs:

- https://github.com/sindresorhus/is-online/pull/43#issuecomment-348755137
- https://github.com/sindresorhus/p-timeout/commit/032432734238d696bb88aeb4a6fc5d4ffc454b17

### PR DESCRIPTION
**Pull request by BendingBender**
Waiting for:
- ~~https://github.com/sindresorhus/p-cancelable/pull/16~~
- ~~https://github.com/sindresorhus/p-some/pull/9~~

---

Fixes #3

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	RecordType,
+	Exclude<keyof RecordType, KeyType>
+>;
+export type Value<T> = T | PromiseLike<T>;
+export type Options<T> = Omit<PSomeOptions<T>, 'count'>;
+export type CancelablePromise<ValueType> = PCancelable<ValueType>;
+
+/**
+ * Wait for any promise to be fulfilled.
+ *
```
Comment (sindresorhus) [inline index.d.ts]: See https://github.com/sindresorhus/meow/pull/116#discussion_r265038725

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * @param input - An `Iterable` collection of promises/values to wait for.
+ * @returns A `Promise` that is fulfilled when any promise from `input` is fulfilled. If all the input promises reject, it will reject with an [`AggregateError`](https://github.com/sindresorhus/aggregate-error) error.
+ */
+export default function pAny<T>(
```
Comment (sindresorhus) [inline index.d.ts]: Any better name than `T`?

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import PCancelable from 'p-cancelable';
+import {Options as PSomeOptions} from 'p-some';
+
+export type Omit<RecordType, KeyType extends keyof RecordType> = Pick<
+	RecordType,
+	Exclude<keyof RecordType, KeyType>
+>;
```
Comment (sindresorhus) [inline index.d.ts]: Could you depend on https://github.com/sindresorhus/type-fest and use https://github.com/sindresorhus/type-fest/blob/aea4f4d646259445b55e08a76c15b831c476dc34/index.d.ts#L65 ?

Comment (sindresorhus) [reply index.d.ts]: Unrelated, but while I got you here. Any feedback on the naming in my version versus yours. `RecordType, KeyType` vs `ObjectType, KeysType`.

File: index.js [hunk 4]
```diff
@@ -1,6 +1,25 @@
 'use strict';
 const pSome = require('p-some');
+const PCancelable = require('p-cancelable');
 
-module.exports = (iterable, opts) => pSome(iterable, Object.assign({}, opts, {count: 1})).then(values => values[0]);
+const pAny = (iterable, options) => {
+	const anyCancelable = pSome(iterable, {...options, count: 1});
+
+	return new PCancelable((resolve, reject, onCancel) => {
```
Comment (sindresorhus) [inline index.js]: I wonder if we could use `PCancelable.fn()` here instead of the constructor.

File: test.js [hunk 5]
```diff
@@ -27,9 +28,37 @@ test('returns the first fulfilled value that passes the filter function', async
 		Promise.resolve('foo'),
 		Promise.resolve('unicorn')
 	];
-	t.deepEqual(await m(f, {filter: val => val === 'unicorn'}), 'unicorn');
+	t.deepEqual(await pAny(f, {filter: val => val === 'unicorn'}), 'unicorn');
+});
+
+test('returns a cancelable promise', t => {
+	const f = [Promise.resolve(1)];
```
Comment (sindresorhus) [inline test.js]:
```suggestion
	const fixture = [Promise.resolve(1)];
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 3 above)
Comment (BendingBender) [reply index.d.ts]: I think `ObjectType` and `KeysType` are clearer.

**Comment by sindresorhus** [comment]
Can you mention in the readme that the returned promise is cancelable (and link to `p-cancelable`)? I think we forgot to do this with `p-some` too.

**Review by sindresorhus — APPROVED** [review]

**Comment by sindresorhus** [comment]
@BendingBender Just wanted to say how incredible thankful I am for all your help with adding TS definitions and your high-quality work. 🙌👍🙏

**Comment by BendingBender** [comment]
You're welcome. Thank you for your great modules!

**Pull request merged**